### PR TITLE
Further functionality for AlgHoms

### PR DESCRIPTION
**Pull request by raulepure**
Added a test for finiteness, added the computation of an inverse map, if existent,
and added the computation of a preimage of an element, if it is in the image.
The preimage function returns, for the moment, a tuple containing the element in
the preimage and the Kernel.
Changed the structure of AlgHoms by storing the groebner data involved in the computations
regarding surjectivity, preimages, inversion and finiteness. Addapted existing code.

### CODE REVIEW

File: src/Rings/mpoly-affine-algebras.jl [hunk 1]
```diff
@@ -29,15 +29,6 @@ end
...
     hilbert_series(A::MPolyQuo)
-
-Given a graded affine algebra $A$ over a field $K$, return a pair $(n,d)$, say, 
-of univariate polynomials in $t$ with integer coefficients
-such that: If $A = R/I$, where $R$ is a multivariate polynomial ring in $n$
-variables over $K$, with positive integer weights $w_1, \dots, w_n$ assigned to the variables,
-and $I$ is a homogeneous ideal of $R$ when we grade $R$ according to the corresponding 
-weighted degree, then $n/d$ represents the Hilbert series of $A$ as a rational function
-with denominator $d = (1-t^{w_1})\cdots (1-t^{w_n})$. See also 
-`hilbert_series_reduced(A::MPolyQuo)`.
```
Comment (fingolfin) [inline src/Rings/mpoly-affine-algebras.jl:40]: Are you removing this on purpose? It was only added today by @wdecker . Maybe a merge or rebase went wrong?

File: src/Rings/mpoly-affine-algebras.jl [hunk 2]
```diff
@@ -246,7 +221,58 @@ end
...
+   psi.kernel = ideal(s, [zero(s)])
+   return psi
+end
+
+function preimage(F::AlgHom, f::Union{MPolyElem, MPolyQuoElem})
+   @assert parent(f) == codomain(F)
+   r = domain(F)
+   s = codomain(F)
+   n = ngens(r)
+   m = ngens(s)
```
Comment (fingolfin) [inline src/Rings/mpoly-affine-algebras.jl:269]: This function is not exercised by any tests, according to codecov.

File: src/Rings/mpoly-affine-algebras.jl (hunk 1 above)
Comment (wdecker) [reply src/Rings/mpoly-affine-algebras.jl:40]: I will take care of this

Comment (raulepure) [reply src/Rings/mpoly-affine-algebras.jl:40]: I can bring it back in. It was probably removed by a rebase

Comment (thofma) [reply src/Rings/mpoly-affine-algebras.jl:40]: @fieker merged this, so does this mean we don't want to add it in?

Comment (wdecker) [reply src/Rings/mpoly-affine-algebras.jl:40]: Solved by merge with 365

**Pull request merged**